## To predict the loaction of a subject based on the call records.

In [1]:
#to find some kind of geogrphical pattern for a person in some are
#source:CDR Data

In [2]:
import pandas as pd
import numpy as np

In [3]:
data = pd.read_csv("tryit.csv", sep = ",")
#print data.head()
#eliminating useless information
data=data.drop(['OTHER','PHONE','TYPE','DURATION','IMEINUMBER','OTYPE','AZIMUTH','OPERATOR','CELLTOWERID','STATE','LAT','LONG'],axis=1)
print data.columns
print data.head()

Index([u'NICKNAME', u'STARTDATE', u'STARTTIME', u'AREADESCRIPTION'], dtype='object')
  NICKNAME   STARTDATE   STARTTIME                  AREADESCRIPTION
0   Jackal  01/01/2014  5:40:06 AM  BSC80:20_MHBD_AnuComplx_WL2033A
1   Jackal  01/01/2014  5:51:43 AM  BSC80:20_MHBD_AnuComplx_WL2033A
2   Jackal  01/01/2014  7:13:29 AM  BSC80:20_MHBD_AnuComplx_WL2033A
3   Jackal  01/01/2014  2:31:08 PM  BSC80:20_MHBD_AnuComplx_WL2033A
4   Jackal  01/01/2014  2:40:11 PM  BSC80:20_MHBD_AnuComplx_WL2033A


In [4]:
#unique candidates in data
print data.NICKNAME.unique()

['Jackal' 'Red Fox' 'King Pin' 'Blackbeard' 'Peacock']


<h1>Models to predict <i>JACKAL's</i> LOCATION</h1>

In [5]:
#print data[['AREADESCRIPTION','LAT','LONG']].drop_duplicates()
mew=data.loc[data['NICKNAME']=='Jackal'].loc[data['AREADESCRIPTION']!='NOT AVAILABLE']
print "distinct Location visited by Jackal::\n",mew.head()

distinct Location visited by Jackal::
  NICKNAME   STARTDATE   STARTTIME                  AREADESCRIPTION
0   Jackal  01/01/2014  5:40:06 AM  BSC80:20_MHBD_AnuComplx_WL2033A
1   Jackal  01/01/2014  5:51:43 AM  BSC80:20_MHBD_AnuComplx_WL2033A
2   Jackal  01/01/2014  7:13:29 AM  BSC80:20_MHBD_AnuComplx_WL2033A
3   Jackal  01/01/2014  2:31:08 PM  BSC80:20_MHBD_AnuComplx_WL2033A
4   Jackal  01/01/2014  2:40:11 PM  BSC80:20_MHBD_AnuComplx_WL2033A


In [6]:
from datetime import date
import calendar
import re
mew['TIME']='__'
mew['DAY']='__'
#print mew
for index, row in mew.iterrows():
    #print row['STARTDATE']
    d,m,y=map(int,row['STARTDATE'].split('/'))
    #print date(y,m,d).weekday()
    row['DAY']=date(y,m,d).weekday()
    g=re.split(':| ',row['STARTTIME'])
    if g[3]=='AM' and g[0]!='12':
        row['TIME']=int(g[0])
        
    elif g[3]=='AM' and g[0]=='12':
        row['TIME']=0
        
    elif g[3]=='PM' and g[0]=='12':
        row['TIME']=12
        
    else:
        row['TIME']=int(g[0])+12        
    


In [7]:
print mew[:15]

   NICKNAME   STARTDATE    STARTTIME                  AREADESCRIPTION TIME DAY
0    Jackal  01/01/2014   5:40:06 AM  BSC80:20_MHBD_AnuComplx_WL2033A    5   2
1    Jackal  01/01/2014   5:51:43 AM  BSC80:20_MHBD_AnuComplx_WL2033A    5   2
2    Jackal  01/01/2014   7:13:29 AM  BSC80:20_MHBD_AnuComplx_WL2033A    7   2
3    Jackal  01/01/2014   2:31:08 PM  BSC80:20_MHBD_AnuComplx_WL2033A   14   2
4    Jackal  01/01/2014   2:40:11 PM  BSC80:20_MHBD_AnuComplx_WL2033A   14   2
5    Jackal  01/01/2014   7:26:21 PM  BSC80:20_MHBD_AnuComplx_WL2033A   19   2
6    Jackal  02/01/2014   6:34:54 AM  BSC80:20_MHBD_AnuComplx_WL2033A    6   3
7    Jackal  02/01/2014   8:00:50 AM  BSC80:20_MHBD_AnuComplx_WL2033A    8   3
8    Jackal  02/01/2014   7:46:02 PM        BSC80:33_Parakala_WL2039A   19   3
9    Jackal  02/01/2014   7:47:57 PM        BSC80:33_Parakala_WL2039A   19   3
10   Jackal  03/01/2014   7:39:20 AM  BSC80:20_MHBD_AnuComplx_WL2033A    7   4
11   Jackal  03/01/2014  10:33:46 AM  BSC80:57_MHBD_

In [8]:
mew=mew.drop(['STARTDATE','STARTTIME','NICKNAME'],axis=1)
mew=mew[['DAY','TIME','AREADESCRIPTION']]

In [9]:
#deleted additional features....... like date,time bcoz we already got genaralised 
print mew.head()

  DAY TIME                  AREADESCRIPTION
0   2    5  BSC80:20_MHBD_AnuComplx_WL2033A
1   2    5  BSC80:20_MHBD_AnuComplx_WL2033A
2   2    7  BSC80:20_MHBD_AnuComplx_WL2033A
3   2   14  BSC80:20_MHBD_AnuComplx_WL2033A
4   2   14  BSC80:20_MHBD_AnuComplx_WL2033A


In [10]:
from IPython.core.display import HTML

<h1>done with preprocessing </h1>

In [11]:
print mew['AREADESCRIPTION'].value_counts()

BSC80:20_MHBD_AnuComplx_WL2033A       324
BSC80:57_MHBD_GovtHsptl_WL2079A        64
BSC80:21_MHBD_TeleExge_WL2003A         20
BSC77:09_TNGOBLDG_WL2114A              14
BSC80:38_Kothaguda_WL2076A             14
BSC80:26_Gudur_WL2028A                 13
BSC80:51_Narsampet2_WL2084A            13
BSC80:28_NarsampetaTE_WL2004A          11
BSC77:18_Tadwai_HW_WL2045A              9
BSC77:20_Medaram_EM_WL2251B             9
BSC77:23_Tirumalagandi_WL2135B          9
BSC77:02_Kambhalapalle_WL2168B          8
BSC80:17_Mulugu_TE_WL2035A              8
BSC80:36_Khanapur_WL2074A               6
BSC01:32_Apparajupally_WL2138B          6
BSC80:08_Pasra_WL2041A                  4
BSC80:56_MondraiH2M_WL2264C             4
BSC01:33_Ootai_WL2198A                  4
BSC77:14_EturnagaramBS_WL2228A          3
BSC77:28_Kesamudram_vlg_WL2136C         2
BSC72:28_Laxettpet_AD2014A              2
BSC80:59_Mallampalle_WL2257B            2
BSC77:25_MDRM_Jakaram_WL2274C           2
BSC77:19_Narlapur_WL2252A         

In [12]:
from sklearn.preprocessing import LabelEncoder

lb_make = LabelEncoder()

mew['AREADESCRIPTION']=lb_make.fit_transform(mew['AREADESCRIPTION'])


dummies = pd.get_dummies(mew["DAY"],prefix="DAY")
mew      = mew.join(dummies)

dummies = pd.get_dummies(mew["TIME"],prefix="TIME")
mew      = mew.join(dummies)


del mew['DAY']
del mew['TIME']

mew_c=mew.copy()
mew_c.drop(['AREADESCRIPTION'],axis=1)


X=mew_c.values
print X
y=mew['AREADESCRIPTION'].values
print y

[[26  0  0 ...,  0  0  0]
 [26  0  0 ...,  0  0  0]
 [26  0  0 ...,  0  0  0]
 ..., 
 [26  0  0 ...,  0  0  0]
 [11  0  0 ...,  0  0  0]
 [41  1  0 ...,  0  0  0]]
[26 26 26 26 26 26 26 26 31 31 26 41 27 41 26 41 41 26  4 34 26 26 26 26 26
 27 26 26 26 26 26 26 26 26  5 30  2 29 26 26 26 27 26 26 26 26 26 26 26 26
 26 26 26 26 26 26 26 26 41 26 27 26 27 26 26 26  0  1 10  8  9  9 26 26 30
 21 23 18 26 26 30 26 26 13 13 13 13 13 13 26 26 26 26 26 26 26 26 27 37 41
 24 18 16 13 27 41  6 26 26 25 25 16 16 16 16 16 26 41 41 13 13 13 13 17 17
 18 18 18 18 18 18 26 22 22  7 13 26 26 26  4 41 41 43 13 32 26 26 26 26 26
 26 26 27 41 26 30 15 15 26 26 26 13 34 29  4 25 25 25 25 41 26 29 26 41 29
 29 26 26 41 27 41 29 29  4 41 26 39 39 39 39 26  4 29 29 29 41 41 36 23 16
 26 12 27 26 26 26 26 26 26 35 41 41 41 39 33 26 26 26 26 26 26 26 26 26 26
 26 26 26 26 26 26 26 26 26 26 26 26 26 26 26 26 34 26 26 11 26 39 26 11 11
 11 11 34 26 26 26 26 26 26 26 26 26 38 38 26 26 26 26 26 26 26 26 26 26 27


<h1>Lets start with Support Vector Machine<h2>


In [13]:
from sklearn import svm
clf=svm.SVC(gamma=0.05, C=15)

In [14]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

In [15]:
from sklearn.model_selection import cross_val_score

scores = cross_val_score(clf,X_train,y_train, cv=27)
print "Cross Validation Score :",scores.mean()*100,"%"

/home/siddharth/miniconda2/lib/python2.7/site-packages/sklearn/model_selection/_split.py:581: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of groups for any class cannot be less than n_splits=27.
  % (min_groups, self.n_splits)), Warning)


Cross Validation Score : 98.2568522401 %


In [16]:
clf.fit(X_train,y_train)
print "Test score,",clf.score(X_test,y_test),"%"

Test score, 0.965517241379 %


<h1>Example Prediction Using SVM</h1>

In [17]:
import numpy
print "example prediction:Jackal's position"
print lb_make.inverse_transform(clf.predict(X_train[0]))
print "\n\n"

print "example prediction:Jackal's position-2"
print lb_make.inverse_transform(clf.predict(X_train[1]))

example prediction:Jackal's position
['BSC72:43_Srirampur_AD2046A']



example prediction:Jackal's position-2
['BSC80:20_MHBD_AnuComplx_WL2033A']


/home/siddharth/miniconda2/lib/python2.7/site-packages/sklearn/utils/validation.py:395: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and will raise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  DeprecationWarning)
/home/siddharth/miniconda2/lib/python2.7/site-packages/sklearn/utils/validation.py:395: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and will raise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  DeprecationWarning)


<h1>Lets try Naive Bayes</h1>

In [18]:
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()

In [19]:
gnb.fit(X_train,y_train)

GaussianNB(priors=None)

In [20]:
print "Prediction Accuracy, ",gnb.score(X_test,y_test)*100,"%"

scores = cross_val_score(gnb,X_train,y_train, cv=27)
print "Cross Validation Score :",scores.mean()*100,"%"

Prediction Accuracy,  93.1034482759 %
Cross Validation Score : 97.716256102 %


<h1>Lets try KNN</h1>

In [21]:
from sklearn.neighbors import KNeighborsClassifier
knn=KNeighborsClassifier(n_neighbors=35)
knn.fit(X_train,y_train)
print "Prediction Accuracy, ",knn.score(X_test,y_test)*100,"%"


scores = cross_val_score(knn,X_train,y_train, cv=27)
print "Cross Validation Score :",scores.mean()*100,"%"

Prediction Accuracy,  74.1379310345 %
Cross Validation Score : 83.9071172172 %


In [22]:
print "example prediction:Jackal's position"
print lb_make.inverse_transform(knn.predict(X_train[6]))
print "\n\n"

print "example prediction:Jackal's position"
print lb_make.inverse_transform(knn.predict(X_test[7]))

example prediction:Jackal's position
['BSC80:20_MHBD_AnuComplx_WL2033A']



example prediction:Jackal's position
['BSC80:20_MHBD_AnuComplx_WL2033A']


/home/siddharth/miniconda2/lib/python2.7/site-packages/sklearn/utils/validation.py:395: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and will raise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  DeprecationWarning)
/home/siddharth/miniconda2/lib/python2.7/site-packages/sklearn/utils/validation.py:395: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and will raise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  DeprecationWarning)


Both KNN and SVM have predicted the same value for trial

In [23]:
from sklearn import linear_model
logistic = linear_model.LogisticRegression(C=5000)
logistic.fit(X_train,y_train)

LogisticRegression(C=5000, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [24]:
print "prediction accuracy",logistic.score(X_test,y_test)*100," %"

scores = cross_val_score(logistic,X_train,y_train, cv=27)
print "Cross Validation Score :",scores.mean()*100,"%"

prediction accuracy 72.4137931034  %
Cross Validation Score : 77.0894402096 %


# FINAL RESULTS

## SVM
1. Test score:96.5%
2. Cross Validation:98.25%

## Naive Bayes
1. Test score:93%
2. Cross Validation:97%

## KNN
1. Test score:74%
2. Cross Validation:83%

## Logistic Regression
1. Test score:72.43%
2. Cross Validation:77.08%

## SVM has the highest test accuracy as well as cross validation Score.
## even Naive Bayes is preforming good